File: sdk/python-sdk/src/browser_agent/runtime.py
from __future__ import annotations

import asyncio
import hashlib
import json
import os
import platform
import shutil
import sys
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Awaitable, Callable

from .errors import BrowserAgentError
from .models import BrowserAgentTask
from .options import ResolvedOptions

def platform_key(system: str | None = None, machine: str | None = None) -> str:
    system = system or sys.platform
    machine = (machine or platform.machine()).lower()
    architecture = {"aarch64": "arm64", "amd64": "x64", "x86_64": "x64"}.get(machine, machine)
    return f"{system}-{architecture}"
def bundled_executable(system: str | None = None, machine: str | None = None) -> Path:
    system = system or sys.platform
    suffix = ".exe" if system == "win32" else ""
    return Path(__file__).parent / "bin" / platform_key(system, machine) / f"browser-agent{suffix}"
def bundled_manifest() -> Path:
    return Path(__file__).parent / "cli-manifest.json"
def verify_bundled_checksum(
    executable: Path,
    key: str,
    manifest: Path | None = None,
) -> None:
    manifest = manifest or bundled_manifest()
    try:
        metadata = json.loads(manifest.read_text(encoding="utf-8"))
        expected = metadata["platforms"][key]["sha256"]
    except (OSError, KeyError, TypeError, json.JSONDecodeError) as error:
        raise BrowserAgentError(
            "CLI_VERSION_INCOMPATIBLE",
            f"Bundled browser-agent checksum metadata is unavailable for {key}.",
        ) from error
    digest = hashlib.sha256()
    with executable.open("rb") as binary:
        for chunk in iter(lambda: binary.read(1024 * 1024), b""):
            digest.update(chunk)
    actual = digest.hexdigest()
    if actual != expected:
        raise BrowserAgentError(
            "CLI_VERSION_INCOMPATIBLE",
            f"Bundled browser-agent checksum verification failed for {key}.",
        )
async def resolve_executable(executable: Path | None = None) -> str:
    bundled = executable is None
    executable = executable or bundled_executable()
    if not executable.is_file() or not os.access(executable, os.X_OK):
        raise BrowserAgentError(
            "CLI_NOT_FOUND",
            f"Bundled browser-agent executable is unavailable for {platform_key()}.",
        )
    if bundled:
        verify_bundled_checksum(executable, platform_key())
    return str(executable)
async def verify_executable(executable: str, timeout: float = 5) -> None:
    try:
        process = await asyncio.create_subprocess_exec(
            executable, "--version-json",
            stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
        )
        stdout, _ = await asyncio.wait_for(process.communicate(), timeout)
    except TimeoutError as error:
        process.kill()
        await process.wait()
        raise BrowserAgentError(
            "CLI_VERSION_INCOMPATIBLE", "Bundled browser-agent version check timed out."
        ) from error
    except OSError as error:
        raise BrowserAgentError(
            "CLI_NOT_FOUND", "Bundled browser-agent executable could not be started."
        ) from error
    try:
        if process.returncode or json.loads(stdout)["rpcProtocolVersion"] != 1:
            raise ValueError
    except (ValueError, KeyError, json.JSONDecodeError) as error:
        raise BrowserAgentError(
            "CLI_VERSION_INCOMPATIBLE",
            "Bundled browser-agent uses an incompatible RPC protocol.",
        ) from error
@dataclass(frozen=True, slots=True)
class ExecutableDependencies:
    resolve: Callable[[], Awaitable[str]]
    verify: Callable[[str], Awaitable[None]]
DEFAULT_EXECUTABLE_DEPENDENCIES = ExecutableDependencies(
    resolve_executable, verify_executable
)
@dataclass(slots=True)
class RuntimeFiles:
    config_path: str
    internal_paths: list[str]
    def cleanup(self) -> None:
        for owned in self.internal_paths:
            shutil.rmtree(owned, ignore_errors=True)
def create_runtime_files(options: ResolvedOptions, tasks: list[BrowserAgentTask]):
    owned: list[str] = []
    try:
        runtime = tempfile.mkdtemp(prefix="browser-agent-sdk-")
        owned.append(runtime)
        os.chmod(runtime, 0o700)
        workspace = options.workspace_directory
        if workspace is None:
            workspace = tempfile.mkdtemp(prefix=".browser-agent-workspace-", dir=os.getcwd())
            owned.append(workspace)
        Path(options.download_directory).mkdir(parents=True, exist_ok=True)
        Path(workspace).mkdir(parents=True, exist_ok=True)
        config_path = str(Path(runtime) / "config.yaml")
        config = {
            "provider": options.provider, "model": options.model,
            "reasoning_effort": options.reasoning_effort,
            **({"endpoint_url": options.endpoint_url} if options.endpoint_url else {}),
            "feature_flags": {"user_takeover_tool": options.user_takeover_tool},
            "headless": options.headless,
            **({"executable_path": options.executable_path} if options.executable_path else {}),
            "download_dir": options.download_directory, "file_workspace_root": workspace,
            "max_steps": options.max_steps, "concurrency": options.concurrency,
            "task_runs": options.runs_per_task, "task_run_retry_count": options.retry_count,
            "validator_lifecycle": {"mode": "terminal", "max_failures": 3},
            "wait_between_tasks_ms": 0, "save_steps_context": True, "save_task_logs": False,
            "step_messages_jsonl_path": str(Path(runtime) / "steps.jsonl"),
            "tasks": [{"task": task.task, **({"url": task.url} if task.url else {})}
                      for task in tasks],
        }
        descriptor = os.open(config_path, os.O_WRONLY | os.O_CREAT, 0o600)
        with os.fdopen(descriptor, "w", encoding="utf-8") as output:
            json.dump(config, output)
        return RuntimeFiles(config_path, owned)
    except BaseException:
        for path in owned:
            shutil.rmtree(path, ignore_errors=True)
        raise

File: sdk/typescript-sdk/src/runtime.ts
import { execFile } from "node:child_process";
import { constants } from "node:fs";
import { access, chmod, mkdir, mkdtemp, rm, writeFile } from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { createRequire } from "node:module";
import { fileURLToPath } from "node:url";
import { promisify } from "node:util";
import { BrowserAgentError } from "./errors.js";
import type { ResolvedOptions } from "./options.js";
import type { BrowserAgentTask } from "./types.js";

const execute = promisify(execFile);
const require = createRequire(import.meta.url);
const PLATFORM_PACKAGES = {
	"darwin-arm64": "crafty-browser-agent-darwin-arm64",
	"darwin-x64": "crafty-browser-agent-darwin-x64",
	"linux-arm64": "crafty-browser-agent-linux-arm64",
	"linux-x64": "crafty-browser-agent-linux-x64",
	"win32-arm64": "crafty-browser-agent-win32-arm64",
	"win32-x64": "crafty-browser-agent-win32-x64",
} as const;

export const platformKey = (
	platform: NodeJS.Platform = process.platform,
	architecture = process.arch,
) => `${platform}-${architecture}`;
export function bundledExecutable(
	platform: NodeJS.Platform = process.platform,
	architecture = process.arch,
): string {
	const suffix = platform === "win32" ? ".exe" : "";
	const key = platformKey(platform, architecture);
	const packageName =
		PLATFORM_PACKAGES[key as keyof typeof PLATFORM_PACKAGES] ??
		`crafty-browser-agent-${key}`;
	try {
		const manifest = require.resolve(`${packageName}/package.json`);
		return path.join(
			path.dirname(manifest),
			"bin",
			`browser-agent${suffix}`,
		);
	} catch {
		const packageRoot = fileURLToPath(new URL("../", import.meta.url));
		return path.resolve(
			packageRoot,
			"..",
			packageName,
			"bin",
			`browser-agent${suffix}`,
		);
	}
}
export async function resolveExecutable(
	executable?: string,
	platform: NodeJS.Platform = process.platform,
	architecture = process.arch,
): Promise<string> {
	const resolved = executable ?? bundledExecutable(platform, architecture);
	try {
		await access(resolved, constants.X_OK);
		return resolved;
	} catch {
		if (!executable) {
			const key = platformKey(platform, architecture);
			const packageName =
				PLATFORM_PACKAGES[key as keyof typeof PLATFORM_PACKAGES];
			throw new BrowserAgentError(
				"CLI_NOT_FOUND",
				packageName
					? `Platform package '${packageName}' is unavailable. Reinstall 'crafty-browser-agent' without omitting optional dependencies.`
					: `Crafty Browser Agent does not provide an executable for ${key}.`,
			);
		}
		throw new BrowserAgentError(
			"CLI_NOT_FOUND",
			`Bundled browser-agent executable is unavailable for ${platformKey()}.`,
		);
	}
}
export async function verifyExecutable(
	executable: string,
	verificationTimeoutMs = 5_000,
): Promise<void> {
	let stdout: string;
	try {
		({ stdout } = await execute(executable, ["--version-json"], {
			encoding: "utf8",
			timeout: verificationTimeoutMs,
		}));
	} catch (cause) {
		const code = (cause as NodeJS.ErrnoException).code;
		const timedOut = Boolean((cause as { killed?: boolean }).killed);
		throw new BrowserAgentError(
			code === "ENOENT" ? "CLI_NOT_FOUND" : "CLI_VERSION_INCOMPATIBLE",
			code === "ENOENT"
				? "Bundled browser-agent executable could not be started."
				: timedOut
					? "Bundled browser-agent version check timed out."
					: "Bundled browser-agent uses an incompatible RPC protocol.",
			{ cause },
		);
	}
	try {
		if (JSON.parse(stdout).rpcProtocolVersion !== 1) throw new Error();
	} catch {
		throw new BrowserAgentError(
			"CLI_VERSION_INCOMPATIBLE",
			"Bundled browser-agent uses an incompatible RPC protocol.",
		);
	}
}
export type ExecutableDependencies = {
	resolve(): Promise<string>;
	verify(executable: string): Promise<void>;
};
export const DEFAULT_EXECUTABLE_DEPENDENCIES = {
	resolve: resolveExecutable,
	verify: verifyExecutable,
} satisfies ExecutableDependencies;

export interface RuntimeFiles {
	configPath: string;
	internalPaths: string[];
	cleanup(): Promise<void>;
}
export async function createRuntimeFiles(
	options: ResolvedOptions,
	tasks: BrowserAgentTask[],
): Promise<RuntimeFiles> {
	const owned: string[] = [];
	try {
		const runtime = await mkdtemp(
			path.join(os.tmpdir(), "browser-agent-sdk-"),
		);
		owned.push(runtime);
		await chmod(runtime, 0o700);
		const workspace =
			options.workspaceDirectory ??
			(await mkdtemp(
				path.join(process.cwd(), ".browser-agent-workspace-"),
			));
		if (!options.workspaceDirectory) owned.push(workspace);
		await mkdir(options.downloadDirectory, { recursive: true });
		await mkdir(workspace, { recursive: true });
		const configPath = path.join(runtime, "config.yaml");
		await writeFile(
			configPath,
			JSON.stringify({
				provider: options.provider,
				model: options.model,
				reasoning_effort: options.reasoningEffort,
				...(options.endpointUrl && {
					endpoint_url: options.endpointUrl,
				}),
				feature_flags: { user_takeover_tool: options.userTakeoverTool },
				headless: options.headless,
				...(options.executablePath && {
					executable_path: options.executablePath,
				}),
				download_dir: options.downloadDirectory,
				file_workspace_root: workspace,
				max_steps: options.maxSteps,
				concurrency: options.concurrency,
				task_runs: options.runsPerTask,
				task_run_retry_count: options.retryCount,
				validator_lifecycle: { mode: "terminal", max_failures: 3 },
				wait_between_tasks_ms: 0,
				save_steps_context: true,
				save_task_logs: false,
				step_messages_jsonl_path: path.join(runtime, "steps.jsonl"),
				tasks: tasks.map(({ task, url }) => ({
					task,
					...(url ? { url } : {}),
				})),
			}),
			{ mode: 0o600 },
		);
		const cleanup = () =>
			Promise.all(
				owned.map((item) => rm(item, { recursive: true, force: true })),
			).then(() => undefined);
		return { configPath, internalPaths: owned, cleanup };
	} catch (error) {
		await Promise.all(
			owned.map((item) => rm(item, { recursive: true, force: true })),
		);
		throw error;
	}
}
